graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java

```diff
@@ -0,0 +1,358 @@
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.graphql;
+
+import static java.util.Collections.emptyList;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.stream.PublisherBasedStreamMessage;
+import com.linecorp.armeria.common.stream.StreamMessage;
+import com.linecorp.armeria.common.websocket.WebSocketWriter;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import graphql.ErrorClassification;
+import graphql.ExecutionInput;
+import graphql.ExecutionResult;
+import graphql.GraphQLError;
+import graphql.language.SourceLocation;
+
+/**
+ * Handles the graphql-ws sub protocol within a web socket.
+ * Handles potentially multiple subscriptions through the same web socket.
+ */
+class GraphqlWSSubProtocol {
+    private static final Logger logger = LoggerFactory.getLogger(GraphqlWSSubProtocol.class);
+    private final GraphqlExecutor graphqlExecutor;
+    private final HashMap<String, ExecutionResultSubscriber> graphqlSubscriptions = new HashMap<>();
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    private static final TypeReference<Map<String, Object>> JSON_MAP =
+            new TypeReference<Map<String, Object>>() {};
+
+    private boolean connectionInitiated;
+
+    private final ServiceRequestContext ctx;
+    private final Map<String, Object> upgradeCtx;
+    private Map<String, Object> connectionCtx = ImmutableMap.of();
+
+    GraphqlWSSubProtocol(ServiceRequestContext ctx, GraphqlExecutor executor) {
+        upgradeCtx = GraphqlServiceContexts.graphqlContext(ctx);
+        graphqlExecutor = executor;
+        this.ctx = ctx;
+    }
+
+    /**
+     * Receives an event and returns a response if one should be sent.
+     */
+    @Nullable
+    public void handle(String event, WebSocketWriter out) {
+        try {
+            final Map<String, Object> eventMap = parseJsonString(event, JSON_MAP);
+            final String type = toStringFromJson("type", eventMap.get("type"));
+            requireNonNull(type, "type");
+            final String id;
+
+            switch (type) {
+                case "connection_init":
+                    if (connectionInitiated) {
+                        // Already initiated, that's an error
+                        throw new Exception("4429: Too many initialisation requests");
```

Comment:
I think this isn't going to work as expected. We need to send the close status explicitly.

Reply:
This exception will close the websocket on the server end. The text is just descriptive.